Allow environment variable definitions in tool XML.

jmchilton opened the pull request:
An `<environment_variables>` top-level XML block can be used to define these.

```
<environment_variables>
    <environment_variable name="FOO">$bar</environment_variable>
    <environment_variable name="FOO2">$bar2</environment_variable
</environment_variables>
```

The expressions are contained in the text of the block.

Implementation Detials:

In order to avoid odd shell expression stuff - the evaluated expressions are written to files and environment variables are read from these files at runtime.

Testing:

A demo tool `test/functional/tools/environment_variables.xml` with a test case that can be executed with the following expression.

```
./run_tests.sh -framework -id environment_variables
```

jmchilton committed:
Allow environment variable definitions in tool XML.

An ``<environment_variables>`` top-level XML block can be used to define these.

```
<environment_variables>
    <environment_variable name="FOO">$bar</environment_variable>
    <environment_variable name="FOO2">$bar2</environment_variable
</environment_variables>
```

The expressions are contained in the text of the block.

Implementation Detials:

In order to avoid odd shell expression stuff - the evaluated expressions are written to files and environment variables are read from these files at runtime.

Testing:

A demo tool ``test/functional/tools/environment_variables.xml`` with a test case that can be executed with the following expression.

```
./run_tests.sh -framework -id environment_variables
```

hexylena commented:
Neat!

nsoranzo commented on lib/galaxy/tools/evaluation.py:
```diff
@@ -501,6 +520,14 @@ def __build_param_file( self ):
+        value = fill_template( template, context=context )
+        f = open( config_filename, "w" )
+        f.write( value )
+        f.close()
```
``` python
with open( config_filename, "w" ) as f:
    f.write( value )
```

nsoranzo commented:
Cool, thanks @jmchilton!

Maybe test/functional/tools/environment_variables.xml can have a more "complicated" template in `<environment_variable>`, e.g. `inttest is $inttest`.

jmchilton committed:
Better file I/O handling for refactored code.

Suggested by @nsoranzo (https://github.com/galaxyproject/galaxy/pull/395#discussion_r33136542).

jmchilton committed:
Add more interesting expression to environment_variables demo tool.

Per suggestion by @nsoranzo (https://github.com/galaxyproject/galaxy/pull/395#issuecomment-114827531).

jmchilton commented:
@nsoranzo Thanks as always for the detailed feedback. I have implemented both changes - specifically I have cleaned up the file handling in that refactored code and added templates demonstrating for and if statements when defining environment variables.

nsoranzo commented:
@jmchilton Amazing!

:+1:

hexylena commented:
:+1:
